### ISSUE TITLE
6.0.0.rc2: The behavior of validation when using `accept_nested_attributes_for` is changed, and ActiveRecord has not working validation for associated models.

### ISSUE THREAD

**Issue by sinsoku** [rails/rails#36822]
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  if ENV["RAILS_BRANCH"]
    gem "rails", github: "rails/rails", branch: ENV["RAILS_BRANCH"]
  else
    gem "rails", "5.2.3"
  end
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table(:articles, force: true)

  create_table :top_pages, force: true do |t|
    t.integer :article_id
  end
end

class Article < ActiveRecord::Base
  has_one :top_page
  accepts_nested_attributes_for :top_page
end

class TopPage < ActiveRecord::Base
  belongs_to :article

  validate -> { errors.add(:base, 'does not publish') }, on: :publish

  def valid?(*args)
    logger.debug "valid?(*args): #{args}"
    super
  end
end

puts
puts "ActiveRecord.version = #{ActiveRecord.version}"
puts

class BugTest < Minitest::Test
  def test_to_valid
    article = Article.create
    top_page = TopPage.create
    article.top_page = top_page

    assert_equal false, article.valid?(:publish)
    assert_includes article.top_page.errors.full_messages, 'does not publish'
  end
end
```

### Expected behavior (5.2.3)

I want the validation of associated models to work as well as v5.2.3.

In v5.2.3, the `valid?` method executed three times, and the last call is passed context.

```
$ ruby issue.rb

ActiveRecord.version = 5.2.3

Run options: --seed 45000

# Running:

D, [2019-07-31T20:09:22.267533 #8937] DEBUG -- :    (0.1ms)  begin transaction
D, [2019-07-31T20:09:22.268150 #8937] DEBUG -- :   Article Create (0.1ms)  INSERT INTO "articles" DEFAULT VALUES
D, [2019-07-31T20:09:22.268418 #8937] DEBUG -- :    (0.1ms)  commit transaction
D, [2019-07-31T20:09:22.270776 #8937] DEBUG -- :    (0.0ms)  begin transaction
D, [2019-07-31T20:09:22.270893 #8937] DEBUG -- : valid?(*args): [nil]
D, [2019-07-31T20:09:22.271439 #8937] DEBUG -- :   TopPage Create (0.1ms)  INSERT INTO "top_pages" DEFAULT VALUES
D, [2019-07-31T20:09:22.271632 #8937] DEBUG -- :    (0.0ms)  commit transaction
D, [2019-07-31T20:09:22.277881 #8937] DEBUG -- :   TopPage Load (0.1ms)  SELECT  "top_pages".* FROM "top_pages" WHERE "top_pages"."article_id" = ? LIMIT ?  [["article_id", 1], ["LIMIT", 1]]
D, [2019-07-31T20:09:22.278065 #8937] DEBUG -- :    (0.0ms)  begin transaction
D, [2019-07-31T20:09:22.279525 #8937] DEBUG -- : valid?(*args): [nil]
D, [2019-07-31T20:09:22.280571 #8937] DEBUG -- :   TopPage Update (0.1ms)  UPDATE "top_pages" SET "article_id" = ? WHERE "top_pages"."id" = ?  [["article_id", 1], ["id", 1]]
D, [2019-07-31T20:09:22.280776 #8937] DEBUG -- :    (0.1ms)  commit transaction
D, [2019-07-31T20:09:22.280895 #8937] DEBUG -- : valid?(*args): [:publish]
.

Finished in 0.023475s, 42.5985 runs/s, 127.7955 assertions/s.
1 runs, 3 assertions, 0 failures, 0 errors, 0 skips
```

### Actual behavior (6.0.0.rc2)

ActiveRecord does not perform validations for associated models.

```
$ RAILS_BRANCH=6-0-stable ruby issue.rb

ActiveRecord.version = 6.0.0.rc2

Run options: --seed 36219

# Running:

D, [2019-07-31T20:11:19.936536 #9056] DEBUG -- :    (0.0ms)  begin transaction
D, [2019-07-31T20:11:19.936708 #9056] DEBUG -- :   Article Create (0.1ms)  INSERT INTO "articles" DEFAULT VALUES
D, [2019-07-31T20:11:19.936930 #9056] DEBUG -- :    (0.0ms)  commit transaction
D, [2019-07-31T20:11:19.938738 #9056] DEBUG -- : valid?(*args): [nil]
D, [2019-07-31T20:11:19.939130 #9056] DEBUG -- :    (0.0ms)  begin transaction
D, [2019-07-31T20:11:19.939283 #9056] DEBUG -- :   TopPage Create (0.1ms)  INSERT INTO "top_pages" DEFAULT VALUES
D, [2019-07-31T20:11:19.939466 #9056] DEBUG -- :    (0.1ms)  commit transaction
D, [2019-07-31T20:11:19.945556 #9056] DEBUG -- :   TopPage Load (0.1ms)  SELECT "top_pages".* FROM "top_pages" WHERE "top_pages"."article_id" = ? LIMIT ?  [["article_id", 1], ["LIMIT", 1]]
D, [2019-07-31T20:11:19.946961 #9056] DEBUG -- : valid?(*args): [nil]
D, [2019-07-31T20:11:19.947547 #9056] DEBUG -- :    (0.1ms)  begin transaction
D, [2019-07-31T20:11:19.947819 #9056] DEBUG -- :   TopPage Update (0.2ms)  UPDATE "top_pages" SET "article_id" = ? WHERE "top_pages"."id" = ?  [["article_id", 1], ["id", 1]]
D, [2019-07-31T20:11:19.947974 #9056] DEBUG -- :    (0.0ms)  commit transaction
F

Failure:
BugTest#test_to_valid [rails_issue.rb:60]:
Expected: false
  Actual: true


rails test rails_issue.rb:55



Finished in 0.027986s, 35.7322 runs/s, 35.7322 assertions/s.
1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

### System configuration
**Rails version**: 5.2.3, 6.0.0.rc2(6-0-stable)

**Ruby version**: ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-darwin18]

**Comment by wjessop** [comment]
0206d9cb7c823fefbf3dfc1d1672be4628986c4e is where the behaviour change was introduced.

**Comment by wjessop** [comment]
The change in behaviour at 0206d9cb7c823fefbf3dfc1d1672be4628986c4e was to prevent an inconsistent behaviour reported in https://github.com/rails/rails/pull/36671, and did so by not validating associations that would not be saved, as it seems that previously saved (and thus irellevant in the current request/execution), but invalid, associations shouldn't prevent the current model from being persisted.

The issue reported here relies on records that would not be saved being validated, which seems completely incompatible with the change I made.

I can't see a middle ground between the two behaviours, so I think this might be one for Rails core to discuss, though I'm happy to talk about it with anyone.

**Comment by alexpooley** [comment]
@sinsoku The behaviour you want appears to be through `validates_associated`. The doc for has_one :validate says "When set to true, validates new objects added to association when saving the parent object. false by default. If you want to ensure associated objects are revalidated on every update, use validates_associated."

The change in https://github.com/rails/rails/commit/0206d9cb7c823fefbf3dfc1d1672be4628986c4e by @wjessop appears to bring single association behaviour in line with collection association behaviour by pushing through the `changed_for_autosave?` logic.

I do feel though that autosave: true as used in the https://github.com/rails/rails/commit/0206d9cb7c823fefbf3dfc1d1672be4628986c4e test suggests save will be called on all children, and hence the validations. The doc for has_one :autosave says it will "always save the associated object" which turns out to hold caveats. Maybe adjust to something like "always save a dirty associated object".

I wrote the test https://github.com/rails/rails/pull/36828 to play with the idea that something that was previously valid is no longer valid. In future when I write code such as this, should I make a PR or just point to my local branch / gist ?

**Comment by sinsoku** [comment]
@alexpooley Thanks for your comment. I tried using  `validates_associated`, but this is different from the intended behavior because it calls `valid?` with an empty array arguments(without context).

I do not have a good way to solve both #36671 and this issue, so I would like to wait for comments from the Rails core team.

As a workaround, I have a way to override the `valid?` method.

```ruby
class Article < ActiveRecord::Base
  has_one :top_page
  accepts_nested_attributes_for :top_page

  def valid?(context = nil)
    if context == :publish
      top_page.valid?(context) && super
    else
      super
    end
  end
end
```

**Comment by alexpooley** [comment]
@sinsoku Yes, sorry! I went down the rabbit hole and never made it back. I think the mixing of save and validation code makes this problematic.

**Comment by kamipo** [comment]
Fixed by #37295.
